internal/peer/redis.go

```diff
@@ -135,15 +135,24 @@ func (p *redisPeers) RegisterUpdatedPeersCallback(cb func()) {
 
 // registerSelf inserts self into the peer list and updates self's entry on a
 // regular basis so it doesn't time out and get removed from the list of peers.
-// If this function stops, this host will get ejected from other's peer lists.
-func (p *redisPeers) registerSelf() {
+// When this function stops, it tries to remove the registered key.
+func (p *redisPeers) registerSelf(done chan struct{}) {
 	tk := time.NewTicker(refreshCacheInterval)
-	for range tk.C {
-		ctx, cancel := context.WithTimeout(context.Background(), p.c.GetPeerTimeout())
-		// every 5 seconds, insert a 30sec timeout record. we ignore the error
-		// here since Register() logs the error for us.
-		p.store.Register(ctx, p.publicAddr, peerEntryTimeout)
-		cancel()
+	for {
+		select {
+		case <-tk.C:
+			ctx, cancel := context.WithTimeout(context.Background(), p.c.GetPeerTimeout())
+			// every interval, insert a timeout record. we ignore the error
+			// here since Register() logs the error for us.
+			p.store.Register(ctx, p.publicAddr, peerEntryTimeout)
+			cancel()
+		case <-done:
+			// unregister ourselves
+			ctx, cancel := context.WithTimeout(context.Background(), p.c.GetPeerTimeout())
+			p.store.Unregister(ctx, p.publicAddr)
```

Comment:
It looks like `Unregister` can return an error with logging it itself, so we care about handling those errors here?